common/src/main/java/com/microsoft/identity/common/internal/ui/embeddedwebview/challengehandlers/PKeyAuthChallengeHandler.java

```diff
@@ -22,21 +22,259 @@
+            throw new ClientException(DEVICE_CERTIFICATE_REQUEST_INVALID, "Context is empty");
+        }
+        if (!headerItems.containsKey(RequestField.CertAuthorities.name())) {
+            throw new ClientException(DEVICE_CERTIFICATE_REQUEST_INVALID, "CertAuthorities is empty");
+        }
+    }
+
+    private IDeviceCertificate getWPJAPIInstance(Class<IDeviceCertificate> certClazz)
+            throws ClientException {
+        final IDeviceCertificate deviceCertProxy;
+        final Constructor<?> constructor;
+        try {
+            constructor = certClazz.getDeclaredConstructor();
+            deviceCertProxy = (IDeviceCertificate) constructor.newInstance((Object[]) null);
+        } catch (NoSuchMethodException | InstantiationException | IllegalAccessException
+                | IllegalArgumentException | InvocationTargetException e) {
+            throw new ClientException(ErrorStrings.DEVICE_CERTIFICATE_API_EXCEPTION,
+                    "WPJ Api constructor is not defined", e);
+        }
+        return deviceCertProxy;
+    }
+
+    class PkeyAuthChallengeRequest {
+        private String mNonce = "";
+
+        private String mContext = "";
+
+        /**
+         * Authorization endpoint will return accepted authorities.
+         * The mCertAuthorities could be empty when either no certificate or no permission for ADFS
+         * service account for the Device container in AD.
+         */
+        private List<String> mCertAuthorities;
+
+        /**
+         * Token endpoint will return thumbprint.
+         */
+        private String mThumbprint = "";
+
+        private String mVersion = null;
+
+        private String mSubmitUrl = "";
+
+        PkeyAuthChallengeRequest(final String nonce,
+                                 final String context,
+                                 final List<String> certAuthorities,
+                                 final String thumbprint,
+                                 final String version,
+                                 final String submitUrl) {
+            mNonce = nonce;
+            mContext = context;
+            mCertAuthorities = certAuthorities;
+            mThumbprint = thumbprint;
+            mVersion = version;
+            mSubmitUrl = submitUrl;
+        }
+    }
+
+    class PkeyAuthChallengeResponse {
+        private String mSubmitUrl;
+
+        private String mAuthorizationHeaderValue;
+
+        String getSubmitUrl() {
+            return mSubmitUrl;
+        }
+
+        void setSubmitUrl(String submitUrl) {
+            mSubmitUrl = submitUrl;
+        }
+
+        String getAuthorizationHeaderValue() {
+            return mAuthorizationHeaderValue;
+        }
+
+        void setAuthorizationHeaderValue(String authorizationHeaderValue) {
+            mAuthorizationHeaderValue = authorizationHeaderValue;
+        }
     }
 
-    void processPkeyAuthChallenge() {
-        //TODO
+    enum RequestField {

```

Comment:
_Maybe_ a top-level `enum` -- maybe just `Strings` (seems to only be used in this way)

Reply:
enum looks much cleaner to me. Not sure the benefits to change them into Strings constants.